apps/server/lib/http/auth.ts

```diff
@@ -0,0 +1,75 @@
+import {
+	JellyfishKernel,
+	SessionContract,
+} from '@balena/jellyfish-types/build/core';
+import bcrypt from 'bcrypt';
+import _ from 'lodash';
+
+// TODO: Move this definition elsewhere, and inject it into the middleware.
+const sessionIdAndTokenValidator =
+	(context, jellyfish: JellyfishKernel) =>
+	async (session: string, sessionToken?: string): Promise<boolean> => {
+		try {
+			const card = await jellyfish.getCardById<SessionContract>(
+				context,
+				jellyfish.sessions!.admin,
+				session,
+			);
+			return bcrypt.compare(sessionToken, card!.data.token!.authentication);
+		} catch (e) {
+			return false;
+		}
+	};
+
+// TODO: Move this definition elsewhere, and inject it into the middleware.
+const sessionIdValidator =
+	(context, jellyfish: JellyfishKernel) =>
+	async (session: string): Promise<boolean> => {
+		try {
+			const card = jellyfish.getCardById<SessionContract>(
+				context,
+				jellyfish.sessions!.admin,
+				session,
+			);
+			if (!card) {
+				return false;
+			}
+			return true;
+		} catch (e) {
+			return false;
+		}
+	};
+
+export const authMiddleware =
+	(jellyfish: JellyfishKernel, options: { guestSession: string }) =>
+	async (request, _response, next) => {
+		const sessionHeader = request.headers['x-session-id'];
+		const authorizationHeader = request.headers['authorization'];
+
+		const authorizationHeaderValue = _.last(_.split(authorizationHeader, ' '));
+
+		request.session = options.guestSession;
+
+		// TODO: take validation function injected via options rather than building them here.
```

Comment:
What would be the benefit of injecting the validation functions via options instead of defining them in this file? If we do end up keeping them in this file, it might be worth exporting them so you can write some simple unit tests for them. Much easier to catch things early with unit tests than wait for integration tests to fail, causing us to spend more time digging into the failure to find the cause.

Reply:
In my opinion, the main benefits of injecting stuff is that it allows us to keep the code cleaner and more maintainable. In this specific scenario, it would result in an auth middleware that doesn't need to be aware of Jellyfish at all. It will simply define a clear interface (e.g. `validateToken`), and a higher-level component will provide it with a function matching that interface. It might also have some second-order benefits, especially if we want to "componentize" the project. Regarding unit tests, I don't see any donwsides. We will still be able to run unit tests for the middleware (possibly with mocking the validation functions) and for the validation functions independently, which I think is even more granular than what we have now.

Reply:
I understand the benefits of injection, but I tend to err on the side of not making something injectable unless I'm convinced that it will actually be useful and/or necessary to do so. One upside of defining things locally and not making them injectable is that the end result is that much simpler. I've worked on projects that were over-engineered with too many injection options that only added confusion without many real world benefits. But like I said in my other comment, if you're convinced this is a definite plus and will be useful/necessary then I'd like to see your implementation and I'll back you up. 👍

Reply:
I understand @joshbwlng. Although I definitely believe JF would benefit from such a change, I think this is something that we should discuss together and considering the whole codebase. That said, I will accept your suggestion so that we can move forward with this PR, and bring the topic for discussion in one of our arch calls.